tfhe/src/high_level_api/integers/signed/base.rs

```diff
@@ -113,6 +114,228 @@ where
+                );
+                crate::FheUint32::new(result)
+            }
+            #[cfg(feature = "gpu")]
+            InternalServerKey::Cuda(_) => {
+                panic!("Cuda devices do not support trailing_ones yet");
+            }
+        })
+    }
+
+    /// Returns the base 2 logarithm of the number, rounded down.
+    ///
+    /// Result has no meaning if self encrypts a value <= 0. See [Self::checked_ilog2]
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::prelude::*;
+    /// use tfhe::{generate_keys, set_server_key, ConfigBuilder, FheBool, FheInt16};
+    ///
+    /// let (client_key, server_key) = generate_keys(ConfigBuilder::default());
+    /// set_server_key(server_key);
+    ///
+    /// let a = FheInt16::encrypt(2i16, &client_key);
+    ///
+    /// let result = a.ilog2();
+    /// let decrypted: u32 = result.decrypt(&client_key);
+    /// assert_eq!(decrypted, 1);
+    /// ```
+    pub fn ilog2(&self) -> crate::FheUint32 {
+        global_state::with_internal_keys(|key| match key {
+            InternalServerKey::Cpu(cpu_key) => {
+                let result = cpu_key.pbs_key().ilog2_parallelized(&self.ciphertext);
+                let result = cpu_key.pbs_key().cast_to_unsigned(
+                    result,
+                    crate::FheUint32Id::num_blocks(cpu_key.pbs_key().message_modulus()),
+                );
+                crate::FheUint32::new(result)
+            }
+            #[cfg(feature = "gpu")]
+            InternalServerKey::Cuda(_) => {
+                panic!("Cuda devices do not support ilog2 yet");
+            }
+        })
+    }
+
+    /// Returns the base 2 logarithm of the number, rounded down.
+    ///
+    /// Also returns a boolean flag that is true if the result is valid (i.e self was > 0)
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::prelude::*;
+    /// use tfhe::{generate_keys, set_server_key, ConfigBuilder, FheBool, FheInt16};
+    ///
+    /// let (client_key, server_key) = generate_keys(ConfigBuilder::default());
+    /// set_server_key(server_key);
+    ///
+    /// let a = FheInt16::encrypt(-1i16, &client_key);
+    ///
+    /// let (result, is_ok) = a.checked_ilog2();
+    ///
+    /// let is_ok = is_ok.decrypt(&client_key);
+    /// assert_eq!(is_ok, false);
+    ///
+    /// let decrypted: u16 = result.decrypt(&client_key);
+    /// assert_eq!(decrypted, 15); // result is meaningless
+    /// ```
+    pub fn checked_ilog2(&self) -> (crate::FheUint32, FheBool) {
+        global_state::with_internal_keys(|key| match key {
+            InternalServerKey::Cpu(cpu_key) => {
+                let (result, is_ok) = cpu_key
+                    .pbs_key()
+                    .checked_ilog2_parallelized(&self.ciphertext);
+                let result = cpu_key.pbs_key().cast_to_unsigned(
+                    result,
+                    crate::FheUint32Id::num_blocks(cpu_key.pbs_key().message_modulus()),
+                );
```

Comment:
we need the cast because we return the same size as the input type by default and we want to stick to the std ilog2 ?

Reply:
By default, the ilog2 resuturns something that has enough blocks to represent the ilog2 value, and to make things easy here we always return a 32 bit int